### PR TITLE
DCOS-14525: refactor containers

### PR DESCRIPTION
**Pull request by Poltergeist**
**Checklist**
- [x] Did you add a JIRA issue in a commit message or as part of the branch name?
- [ ] Did you add new unit tests?
- [ ] Did you add new integration tests?
- [ ] If this is a regression, did you write a test to catch this in the future?

closes DCOS-14525

### CODE REVIEW

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+    if (field === 'healthCheck') {
+      if (this.healthCheckState[index] == null) {
+        this.healthCheckState[index] = {};
+      }
+
+      newState[index].healthCheck = multiContainerHealthFormReducer.call(
+        this.healthCheckState[index],
+        newState[index].healthCheck,
+        {type, path: path.slice(3), value}
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js]: Why only 3?

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    if (type === SET && field === 'resources') {
+      // Do not parse numbers, as user might be in the middle of
+      // entering a number. 0.0 would then equal 0, yielding it impossible to
+      // enter, e.g. 0.001
+      newState[index].resources = containerReducer.call(
+        this.cache[index],
+        newState[index].resources,
+        {type, value, path: path.slice(2)}
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js]: Why only 2?

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Endpoints.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    const [_, index, field, secondIndex, name, subField] = path;
+
+    if (field !== 'endpoints') {
+      return state;
+    }
+
+    switch (type) {
+      case ADD_ITEM:
+        const endpointDefinition = Object.assign(
+          {}, defaultEndpointsFieldValues);
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Endpoints.js]: #nit weird position of the closing parenthesis. Not important

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Endpoints.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  },
+  servicePort: null,
+  vip: null
+};
+
+module.exports = {
+  FormReducer(state = [], {type, path = [], value}) {
+    let newState = [].concat(state);
+
+    // eslint-disable-next-line no-unused-vars
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Endpoints.js]: Unused variables that start with underscore are ignored by the linter

File: plugins/services/src/js/reducers/serviceForm/FormReducers/MultiContainerArtifacts.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  SET,
+  ADD_ITEM,
+  REMOVE_ITEM
+} from '../../../../../../../src/js/constants/TransactionTypes';
+
+module.exports = {
+  FormReducer(state = [], {type, path = [], value}) {
+    let newState = [].concat(state);
+
+    // eslint-disable-next-line no-unused-vars
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/MultiContainerArtifacts.js]: I'll let you find the rest of these

File: plugins/services/src/js/reducers/serviceForm/FormReducers/__tests__/Containers-test.js [hunk 6]
```diff
@@ -0,0 +1,898 @@
...
+          ]);
+        });
+
+        it('should have one endpoint with name', function () {
+          let batch = new Batch();
+
+          batch = batch.add(new Transaction(['containers'], 0, ADD_ITEM));
+
+          batch =
+              batch.add(new Transaction([
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/__tests__/Containers-test.js]: Could we clean up some of these weird indentations and newlines? This is not really nice to look at

File: plugins/services/src/js/reducers/serviceForm/FormReducers/__tests__/Containers-test.js [hunk 7]
```diff
@@ -0,0 +1,898 @@
...
+                'containers',
+                0,
+                'endpoints',
+                0,
+                'name'
+              ], 'foo'));
+
+          expect(batch.reduce(Containers.FormReducer.bind({})))
+          .toEqual([
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/__tests__/Containers-test.js]: And this indentation is off too. I know our linter should catch this. I do not know why it is not

File: plugins/services/src/js/reducers/serviceForm/FormReducers/__tests__/MultiContainerArtifacts-test.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        new Transaction(['containers', 0, 'artifacts'], 1, ADD_ITEM),
+        new Transaction([
+          'containers',
+          0,
+          'artifacts',
+          1,
+          'uri'
+        ], 'http://mesosphere.com', SET),
+        new Transaction(['containers', 0, 'artifacts'], 2, ADD_ITEM),
+        new Transaction(['containers', 0, 'artifacts'], 3, ADD_ITEM)
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/__tests__/MultiContainerArtifacts-test.js]: What about testing for removal of artifacts?

File: plugins/services/src/js/reducers/serviceForm/JSONReducers/__tests__/MultiContainerArtifacts-test.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        ], 'http://mesosphere.io', SET),
+        new Transaction(['containers', 0, 'artifacts'], 1, ADD_ITEM),
+        new Transaction([
+          'containers',
+          0,
+          'artifacts',
+          1,
+          'uri'
+        ], 'http://mesosphere.com', SET),
+        new Transaction(['containers', 0, 'artifacts'], 2, ADD_ITEM),
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/JSONReducers/__tests__/MultiContainerArtifacts-test.js]: Again, let's test for removal as well

File: plugins/services/src/js/reducers/serviceForm/JSONReducers/Containers.js [hunk 10]
```diff
@@ -286,6 +264,7 @@ function containersParser(state) {
 
 module.exports = {
   JSONReducer(state = [], {type, path = [], value}) {
+    // eslint-disable-next-line no-unused-vars
```
Comment (Poltergeist) [inline plugins/services/src/js/reducers/serviceForm/JSONReducers/Containers.js]: changed that nice I didn't know that this was possible.

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+const containerReducer = combineReducers({
+  cpus: simpleReducer('resources.cpus'),
+  mem: simpleReducer('resources.mem'),
+  disk: simpleReducer('resources.disk')
+});
+
+module.exports = {
+  FormReducer(state, {type, path = [], value}) {
+    // eslint-disable-next-line no-unused-vars
```
Comment (Poltergeist) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js]: changed that

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js (hunk 1 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js]: Don't know I copied the code basically will pass the complete path here.

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js (hunk 2 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js]: Don't know I copied the code basically will pass the complete path here.

File: plugins/services/src/js/reducers/serviceForm/FormReducers/__tests__/MultiContainerArtifacts-test.js (hunk 8 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/__tests__/MultiContainerArtifacts-test.js]: added a test for removal

File: plugins/services/src/js/reducers/serviceForm/JSONReducers/__tests__/MultiContainerArtifacts-test.js (hunk 9 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/serviceForm/JSONReducers/__tests__/MultiContainerArtifacts-test.js]: added a test for removal

**Comment by Poltergeist** [comment]
@mlunoe all Comments should be addressed.

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js (hunk 1 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js]: 🤔 could we get a comment in from whoever implemented this in the first place?

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js (hunk 2 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js]: 🤔 could we get a comment in from whoever implemented this in the first place?

**Comment by mlunoe** [comment]
Aside from these two, which I think would be nice, I think this is looking good:
https://github.com/dcos/dcos-ui/pull/2061#discussion_r111510994
https://github.com/dcos/dcos-ui/pull/2061#discussion_r111510996

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js (hunk 1 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Containers.js]: Weren't you going to pass in the names in a new array?
Is it `[index, field, subField]`?

**Comment by mlunoe** [comment]
Dangit, konfliktz!

**Comment by nLight** [comment]
No idea if we ever able to rescue that one, but the general concern is that we should 
1. *Move* the code around
2. Then make it better

one step at a time

**Comment by nLight** [comment]
I'll try to cherrypick commits and replay it on top of #2141

**Comment by orlandohohmeier** [comment]
@nLight @Poltergeist any chance we can at least rescue parts of this one? Otherwise let's just close this one and redo.

**Comment by nLight** [comment]
Closing in favor of https://github.com/dcos/dcos-ui/pull/2190

**Pull request closed without merging**